command/windows_service_install.go

```diff
@@ -0,0 +1,343 @@
+	}
+	if opts.dataDir, err = c.winPaths.Expand(opts.dataDir); err != nil {
+		return fmt.Errorf("cannot generate data path - %s", err)
+	}
+
+	// Ensure directories exist
+	if err = c.winPaths.CreateDirectory(opts.configDir, true); err != nil {
+		return fmt.Errorf("cannot create configuration directory - %s", err)
+	}
+
+	if err = c.winPaths.CreateDirectory(opts.dataDir, true); err != nil {
+		return fmt.Errorf("cannot create data directory - %s", err)
+	}
+
+	// Check if any configuration files exist
+	matches, _ := filepath.Glob(filepath.Join(opts.configDir, "*"))
+	if len(matches) < 1 {
+		f, err := os.Create(filepath.Join(opts.configDir, "config.hcl"))
+		if err != nil {
+			return fmt.Errorf("could not create default configuration file - %s", err)
+		}
+		fmt.Fprintf(f, strings.TrimSpace(`
+# Full configuration options can be found at https://developer.hashicorp.com/nomad/docs/configuration
+
+data_dir  = "%s"
+bind_addr = "0.0.0.0"
+
+server {
+  # license_path is required for Nomad Enterprise as of Nomad v1.1.1+
+  #license_path = "%s\license.hclic"
+  enabled          = true
+  bootstrap_expect = 1
+}
+
+client {
+  enabled = true
+  servers = ["127.0.0.1"]
+}
+
+log_level = "WARN"
+eventlog {
+  enabled = true
+  level   = "ERROR"
+}
+`), strings.ReplaceAll(opts.dataDir, `\`, `\\`), strings.ReplaceAll(opts.configDir, `\`, `\\`))
+		f.Close()
+		c.Ui.Output(fmt.Sprintf("  Added initial configuration file: %s", f.Name()))
+	}
+
+	return nil
+}
+
+func (c *WindowsServiceInstallCommand) binaryInstall(opts *windowsInstallOpts) error {
+	// Get the path to the currently executing nomad. This
+	// will be installed for the service to call.
+	exePath, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("cannot detect current nomad path - %s", err)
+	}
+
+	// Build the needed paths
+	if opts.installDir == "" {
+		opts.installDir = winsvc.WINDOWS_INSTALL_BIN_DIRECTORY
+	}
+	opts.installDir, err = c.winPaths.Expand(opts.installDir)
+	if err != nil {
+		return fmt.Errorf("cannot generate binary install path - %s", err)
+	}
+	opts.binaryPath = filepath.Join(opts.installDir, "nomad.exe")
+
+	// Ensure the install directory exists
+	if err = c.winPaths.CreateDirectory(opts.installDir, false); err != nil {
+		return fmt.Errorf("could not create binary install directory - %s", err)
+	}
+
+	// Create a new copy of the current binary to install
+	exeFile, err := os.Open(exePath)
+	if err != nil {
+		return fmt.Errorf("cannot open current nomad path for install - %s", err)
+	}
+	defer exeFile.Close()
+
+	// Copy into a temporary file which can then be moved
+	// into the correct location.
+	dstFile, err := os.CreateTemp(os.TempDir(), "nomad*")
+	if err != nil {
+		return fmt.Errorf("cannot create copy - %s", err)
+	}
+	defer dstFile.Close()
+
+	if _, err = io.Copy(dstFile, exeFile); err != nil {
+		return fmt.Errorf("cannot write nomad binary for install - %s", err)
+	}
+	dstFile.Close()
```

Comment:
I'm sure this is fine, but the defer Close followed so quickly by another non-deferred Close feels funny.

Reply:
The deferred close is there for the copy if it errors out. The followup close just to ensure everything is tidied up (and flushed) before moving the file.